File: code/Java/src/test/java/MarsRoverReceiverShould.java
import Infrastructure.Bus.IMessageReceivedBus;
import Infrastructure.SpaceComm.MarsRoverReceiver;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

public class MarsRoverReceiverShould {
    MarsRoverReceiver marsRoverReceiver;
    private IMessageReceivedBus mockServiceBus;

    private static Stream<Arguments> valuesProvider() {
        return Stream.of(
                Arguments.of(
                        new String[]{"X2", "Y5", "DN", "M5", "1F", "2L", "3F", "4R", "5F"},
                        "100 100\n2 5 N\nFLFRF"),
                Arguments.of(
                        new String[]{"X2", "Y5", "DN", "M4",  "3F","2L","1F",  "4R"},
                        "100 100\n2 5 N\nFLFR")
        );
    }

    @ParameterizedTest
    @MethodSource("valuesProvider")
    void rebuildMessageCorrectly(String[] packages, String rebuiltMessage) {
        marsRoverReceiver = new MarsRoverReceiver();
        mockServiceBus = mock(IMessageReceivedBus.class);
        marsRoverReceiver.writesTo(mockServiceBus);

        for (String datagram : packages) {
            marsRoverReceiver.received(datagram);
        }

        verify(mockServiceBus).NotifyMessageReceived(rebuiltMessage);
    }

    @Test
    void notifyErrorCorrectly() throws InterruptedException {
        int sleepTime = 3100;
        marsRoverReceiver = new MarsRoverReceiver();
        mockServiceBus = mock(IMessageReceivedBus.class);
        marsRoverReceiver.writesTo(mockServiceBus);

        String[] packages = new String[]{"X2", "Y5", "DN", "M5", "2L", "3F", "4R", "5F"};
        for (String datagram : packages) {
            marsRoverReceiver.received(datagram);
        }

        Thread.sleep(sleepTime);

        verify(mockServiceBus).NotifyError();
    }
}

File: code/CSharp/ConnascentMarsRover/Tests/MarsRoverReceiverShould.cs
using System.Collections;
using System.Threading;
using NSubstitute;
using NUnit.Framework;
using Source.Infrastructure.Bus;
using Source.Infrastructure.SpaceComm;

namespace Tests
{
    [TestFixture]
    public class MarsRoverReceiverShould
    {
        private MarsRoverReceiver _marsRoverReceiver;

        private class MarsRoverReceiverTestCases
        {
            public static IEnumerable TestCases
            {
                get
                {
                    yield return new TestCaseData(
                        new[] {"X2", "Y5", "DN", "M5", "1F", "2L", "3F", "4R", "5F"},
                        "100 100\n2 5 N\nFLFRF");

                    yield return new TestCaseData(
                        new[] {"X2", "Y5", "DN", "M4", "3F", "2L", "1F", "4R"},
                        "100 100\n2 5 N\nFLFR");
                }
            }
        }

        [TestCaseSource(typeof(MarsRoverReceiverTestCases), nameof(MarsRoverReceiverTestCases.TestCases))]
        public void RebuildMessageCorrectly(string[] packets, string rebuiltMessage)
        {
            const int maxDelay = 100;
            _marsRoverReceiver = new MarsRoverReceiver(maxDelay);
            var mockServiceBus = Substitute.For<IMessageReceivedBus>();
            _marsRoverReceiver.WritesTo(mockServiceBus);

            foreach (var packet in packets)
            {
                _marsRoverReceiver.Received(packet);
            }

            mockServiceBus.Received().NotifyMessageReceived(rebuiltMessage);
        }

        [Test]
        public void NotifyErrorCorrectly()
        {
            const int maxDelay = 100;
            _marsRoverReceiver = new MarsRoverReceiver(maxDelay);
            var mockServiceBus = Substitute.For<IMessageReceivedBus>();
            _marsRoverReceiver.WritesTo(mockServiceBus);

            var packets = new[] {"X2", "Y5", "DN", "M5", "2L", "3F", "4R", "5F"};
            foreach (var packet in packets)
            {
                _marsRoverReceiver.Received(packet);
            }

            Thread.Sleep(maxDelay + 30);

            mockServiceBus.Received().NotifyError();
        }
    }
}